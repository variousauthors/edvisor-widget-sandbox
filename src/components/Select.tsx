import * as React from 'react';

interface SelectProps {
  data: any[],
  onChange: (value: string) => void,
  defaultText?: string,
  defaultValue?: string | undefined,
  valueKey?: string,
  nameKey?: string,
}

export default function Select ({ data, onChange, defaultText = "", defaultValue = undefined, valueKey = "value", nameKey = "name"}: SelectProps) {

  let optionElements = data.map((opt: any, index: number) => {

    return (
      <option key={ index } value={ opt[valueKey] } >{ opt[nameKey] }</option>
    )
  })

  if (defaultText !== "") {
    optionElements.unshift((
      <option key={ optionElements.length } value={ defaultValue } >{ defaultText }</option>
    ))
  }

  return (
    <select defaultValue={ defaultValue } onChange={ (e) => { onChange(e.target.value); } }>
      { optionElements }
    </select>
  )
}

