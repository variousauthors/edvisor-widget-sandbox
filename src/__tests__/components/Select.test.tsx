import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import Select from '../../components/Select'

let nothing = (value) => {};
let options = [
  { value: "1", name: "a" },
  { value: "2", name: "b" },
]

describe('<Select />', () => {
	it('renders', () => {

		const tree = renderer.create(<Select data={ options } onChange={ nothing } />).toJSON()
		expect(tree).toMatchSnapshot()
	})

  it('renders the correct number of <option /> tags', () => {

    const wrapper = shallow(<Select data={ options } onChange={ nothing } />);
    expect(wrapper.find('option')).toHaveLength(2);
    expect(wrapper.find('option').first().text()).toBe("a");
  });

  it('renders 1 additinal <option/> tag when defaultText is given', () => {

		const tree = renderer.create(<Select defaultText="Default Option" data={ options } onChange={ nothing } />).toJSON()
    const wrapper = shallow(<Select defaultText="Default Option" data={ options } onChange={ nothing } />);

    expect(wrapper.find('option')).toHaveLength(3);
    expect(wrapper.find('option').first().text()).toBe("Default Option");
		expect(tree).toMatchSnapshot()
  });

  it('uses valueKey and valueName to create options from the given data', () => {
    let weirdOptions = [
      { bob: 1, jill: "a", wat: 3 }
      { bob: 2, jill: "b", wat: 3 }
    ];

    const wrapper = shallow(<Select valueKey="bob" nameKey="jill" data={ weirdOptions } onChange={ nothing } />);

    expect(wrapper.find('option')).toHaveLength(2);
    expect(wrapper.find('option').first().text()).toBe("a");
  });

})

