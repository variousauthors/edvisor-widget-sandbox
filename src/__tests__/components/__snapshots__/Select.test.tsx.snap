// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders 1`] = `
<select
  defaultValue={undefined}
  onChange={[Function]}
>
  <option
    value="1"
  >
    a
  </option>
  <option
    value="2"
  >
    b
  </option>
</select>
`;

exports[`<Select /> renders 1 additinal <option/> tag when defaultText is given 1`] = `
<select
  defaultValue="-1"
  onChange={[Function]}
>
  <option
    value="-1"
  >
    Default Option
  </option>
  <option
    value="1"
  >
    a
  </option>
  <option
    value="2"
  >
    b
  </option>
</select>
`;
